pkg/controller/enterprisesearch/config.go

```diff
@@ -99,7 +137,32 @@ func newConfig(c k8s.Client, ents entsv1beta1.EnterpriseSearch) (*settings.Canon
 	}
 
 	// merge with user settings last so they take precedence
-	if err := cfg.MergeWith(associationCfg, tlsConfig(ents), userProvidedCfg); err != nil {
+	if err := cfg.MergeWith(secretCfg, associationCfg, tlsConfig(ents), userProvidedCfg); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// getExistingConfig retrieves the canonical config, if one exists
+func getExistingConfig(client k8s.Client, ents entsv1beta1.EnterpriseSearch) (*settings.CanonicalConfig, error) {
+	var secret corev1.Secret
+	key := types.NamespacedName{
+		Namespace: ents.Namespace,
+		Name:      name.Config(ents.Name),
+	}
+	err := client.Get(key, &secret)
+	if err != nil && apierrors.IsNotFound(err) {
+		log.V(1).Info("Enterprise Search config secret does not exist", "namespace", ents.Namespace, "ents_name", ents.Name)
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	rawCfg, exists := secret.Data[ConfigFilename]
+	if !exists {
+		return nil, fmt.Errorf("%s:%s: config secret %v exists but config file key %s is missing", ents.Namespace, ents.Name, key, ConfigFilename)
```

Comment:
Should we return an error here?
If I manually remove the secret data, I think it can lead to the reconciliation being stuck forever. Let's maybe log a warning and return an empty config but no error.